src/aotextarea.cpp

```diff
@@ -81,3 +80,44 @@ void AOTextArea::auto_scroll(QTextCursor old_cursor, int old_scrollbar_value,
     this->verticalScrollBar()->setValue(this->verticalScrollBar()->maximum());
   }
 }
+
+std::string AOTextArea::closetags(std::string html)
+{
+  std::regex opened_regex(R"(<([a-z]+)(?: .*)?(?<![/|/ ])>)",
+                          std::regex_constants::icase);
+  std::regex closed_regex(R"(</([a-z]+)>)", std::regex_constants::icase);
+
+  std::vector<std::string> openedtags;
+  std::vector<std::string> closedtags;
+  std::smatch match;
+
+  std::string::const_iterator search_start(html.cbegin());
+  while (std::regex_search(search_start, html.cend(), match, opened_regex)) {
+    openedtags.push_back(match[1]);
+    search_start = match.suffix().first;
+  }
+
+  search_start = html.cbegin();
+  while (std::regex_search(search_start, html.cend(), match, closed_regex)) {
+    closedtags.push_back(match[1]);
+    search_start = match.suffix().first;
+  }
+
+  size_t len_opened = openedtags.size();
+
+  if (closedtags.size() == len_opened) {
+    return html;
+  }
+
+  std::reverse(openedtags.begin(), openedtags.end());
+  for (size_t i = 0; i < len_opened; i++) {
+    auto it = std::find(closedtags.begin(), closedtags.end(), openedtags[i]);
+    if (it == closedtags.end()) {
+      html += "</" + openedtags[i] + ">";
+    }
+    else {
+      closedtags.erase(it);
+    }
+  }
+  return html;
+}
```

Comment:
can you rewrite this to use the Qt functions rather than std:: ones? we must be breaking some kind of convention somewhere by doing it this way lol